cumulus/pallets/xcmp-queue/src/lib.rs

```diff
@@ -722,24 +715,55 @@ impl<T: Config> Pallet<T> {
 			T::WeightInfo::take_first_concatenated_xcm(xcm_data.len() as u32) - base_weight;
 		meter.consume(extra_weight);
 
-		let xcm = Some(BoundedSlice::try_from(xcm_data).map_err(|error| {
+		let xcm = BoundedSlice::try_from(xcm_data).map_err(|error| {
 			tracing::error!(
 				target: LOG_TARGET,
 				?error,
 				"Failed to take XCM after decoding: message is too long"
 			);
 			()
-		})?);
+		})?;
 
 		Ok(xcm)
 	}
 
-	/// Split concatenated encoded `VersionedXcm`s or `MaybeDoubleEncodedVersionedXcm`s into
-	/// batches.
+	/// Split concatenated opaque `VersionedXcm`s into individual items.
+	///
+	/// This method is not benchmarked because it's almost free.
+	pub(crate) fn take_first_concatenated_opaque_xcm<'a>(
+		data: &mut &'a [u8],
+	) -> Result<BoundedSlice<'a, u8, MaxXcmpMessageLenOf<T>>, ()> {
+		let xcm_len = Compact::<u32>::decode(data).map_err(|error| {
+			tracing::debug!(target: LOG_TARGET, ?error, "Failed to decode opaque XCM length");
+			()
+		})?;
+		let (xcm_data, remaining_data) = match data.split_at_checked(xcm_len.0 as usize) {
+			Some((xcm_data, remaining_data)) => (xcm_data, remaining_data),
+			None => {
+				tracing::debug!(target: LOG_TARGET, "Wrong opaque XCM length");
+				return Err(())
+			},
+		};
+		*data = remaining_data;
```

Comment:
Not sure how to handle the error case with this assignment. I guess errors should be muss less likely here since its double-encoded already and an attacker cannot mess up the encoding.

Otherwise I would do the assignment after the error return to give the caller the possibility to explicitly disregard the bad message.

Reply:
You mean after the error return above ? The problem is that there we don't have a `remaining_data` because `split_at_checked` failed. But in case of an error, we skip the remaining of the XCMP page anyway. So we should be ok.